src/Raven.Server/Documents/ShardedTcpHandlers/ShardedSubscriptionConnection.cs

```diff
@@ -0,0 +1,353 @@
+                            continue;
+
+                        exceptions.Add(shard, shardHolder.PullingTask.Exception);
+
+                        if (stopping)
+                            continue;
+
+                        if (shardHolder.LastErrorDateTime.HasValue == false)
+                        {
+                            shardHolder.LastErrorDateTime = DateTime.Now;
+                            shardsToReconnect.Add(shard);
+                        }
+                        else if (DateTime.Now - shardHolder.LastErrorDateTime.Value < _rejectedReconnectTimeout)
+                        {
+                            shardsToReconnect.Add(shard);
+                        }
+                        else
+                        {
+                            // we are stopping this subscription
+                            stopping = true;
+                        }
+                    }
+
+                    if (stopping)
+                        throw new AggregateException($"Stopping sharded subscription '{_options.SubscriptionName}' with id '{SubscriptionId}' for database '{TcpConnection.ShardedContext.DatabaseName}' because shard {string.Join(", ", exceptions.Keys)} workers failed.", exceptions.Values);
+
+                    // try reconnect faulted workers, this will happen every 3 sec
+                    foreach (var shard in shardsToReconnect)
+                    {
+                        CancellationTokenSource.Token.ThrowIfCancellationRequested();
+
+                        if (_shardWorkers.ContainsKey(shard) == false)
+                            continue;
+
+                        using (var old = _shardWorkers[shard].Worker)
+                        {
+                            _shardWorkers[shard].Worker = new ShardedSubscriptionWorker(_options, shard, _shardWorkers[shard].RequestExecutor, SendConfirmToClientUnderLock, ReportExceptionToAndStopProcessing, WriteBatchToClientAndAck);
+                            var t = _shardWorkers[shard].Worker.RunInternal(CancellationTokenSource.Token);
+                            _shardWorkers[shard].PullingTask = t;
+                        }
+                    }
+                }
+
+                CancellationTokenSource.Token.ThrowIfCancellationRequested();
+            }
+        }
+
+        internal async Task SendConfirmToClientUnderLock()
+        {
+            Debug.Assert(_lock.CurrentCount == 0, "SendConfirmToClientUnderLock is NOT under lock");
+
+            try
+            {
+                AddToStatusDescription($"Shard subscription connection '{SubscriptionId}' send confirm to client '{TcpConnection.TcpClient.Client.RemoteEndPoint}'.");
+                await WriteJsonAsync(new DynamicJsonValue { [nameof(SubscriptionConnectionServerMessage.Type)] = nameof(SubscriptionConnectionServerMessage.MessageType.Confirm) }, TcpConnection);
+            }
+            finally
+            {
+                _lock.Release();
```

Comment:
This is really scary. We accept the lock, and then release it? That isn't usually how we do this.
We can assert that we have the lock, but the caller is the one that should release it.